### PR TITLE
Support multiple read-only activations Downstairs

### PR DESCRIPTION
**Pull request by jmpesp**
Previously, the Downstairs would only accept one active Upstairs at a
time. This commit adds support Downstairs for multiple read-only
Upstairs connecting, and sending reads.

This was done by storing a Work struct per Upstairs connection, and
getting the Downstairs to select the correct one based on an
UpstairsConnection argument.

The biggest change in this commit is to `promote_to_active`, which has
different behaviour now depending on if the Downstairs is in read-only
or read-write mode. The read-write path should be functionally the same
as before this commit. Tests were also added to validate
`promote_to_active`'s new behaviour.

### CODE REVIEW

File: downstairs/src/lib.rs [hunk 1]
```diff
@@ -1499,106 +1554,231 @@ impl Downstairs {
...
-            work.active = HashMap::new();
-        }
+            Ok(())
+        } else {
+            // Only one active read-write session is allowed. Kick out the
+            // currently active Upstairs session if one exists.
+            let currently_active_upstairs_uuids: Vec<Uuid> =
+                self.active_upstairs.keys().copied().collect();
+
+            match currently_active_upstairs_uuids.len() {
```
Comment (leftwo) [inline downstairs/src/lib.rs:1635]: Similar locking question here, thought it seems like we are okay.  The action of getting
currently_active_upstairs_uuids, that is atomic?  Meaning, could two connections both call that `self.active_upstairs.keys().copied().collect();` and come back with `0`?

Actually, I think it could, as the time of check is different that the time of add below, so unless we
lock around this, I suspect it's possible for two connections to sneak in here.

File: downstairs/src/lib.rs [hunk 2]
```diff
@@ -3228,4 +3431,318 @@ mod test {
...
+        ds.promote_to_active(upstairs_connection_1, tx1).await?;
+
+        assert_eq!(ds.active_upstairs().len(), 1);
+        assert!(matches!(rx1.try_recv().err().unwrap(), TryRecvError::Empty));
+        assert!(matches!(rx2.try_recv().err().unwrap(), TryRecvError::Empty));
+
+        ds.promote_to_active(upstairs_connection_2, tx2).await?;
+        assert_eq!(rx1.try_recv().unwrap(), upstairs_connection_2);
+        assert!(matches!(rx2.try_recv().unwrap_err(), TryRecvError::Empty));
+
```
Comment (leftwo) [inline downstairs/src/lib.rs:3542]: Could we also test `is_active` in here?  Making sure `upstairs_connection_1` is not active and `upstairs_connection_2` is?

File: integration_tests/src/lib.rs [hunk 3]
```diff
@@ -1432,4 +1432,79 @@ mod test {
 
         Ok(())
     }
+
+    // Test that multiple upstairs can connect to a single read-only downstairs
+    #[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+    async fn integration_test_multi_read_only() -> Result<()> {
+        const BLOCK_SIZE: usize = 512;
+
+        let mut opts = three_downstairs(54040, 54041, 54042, true).unwrap();
```
Comment (leftwo) [inline integration_tests/src/lib.rs]: These port numbers are already in use on line 907 and can cause the test to hang.

I wish I had some better way of allocating port numbers, as I do this to myself all the time.

Also, I was trying to group the "volume" tests into one section (ending at line ~986), and having their ports start with 54xxx and the guest section of tests (which is where this is) all start with 55xxx.

**Review by leftwo — COMMENTED** [review]
Some thoughts and locking concerns below, take a read and let me know what you think.

It may be worth making a higher level test that starts two upstairs talking to the same downstairs RO, then has the downstairs restart while the upstairs continue to send IO.  That sort of test (test_reconnect.sh) found a bunch of edge cases for me when doing the reconnect work.  Maybe make an issue for that as I need to update the dsc program a tiny bit more to allow it to do the work instead of using `downstairs_daemon.sh`.

File: downstairs/src/lib.rs [hunk 4]
```diff
@@ -1168,7 +1188,7 @@ where
              * trigger once then never again.
              */
             _ = sleep_until(lossy_interval), if lossy => {
-                //_show_work(&ds).await;
+                //show_work(&ds).await;
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply downstairs/src/lib.rs:1192]: done in 59de8ef

File: downstairs/src/lib.rs [hunk 5]
```diff
@@ -1349,25 +1373,31 @@ impl Downstairs {
...
+
+        let active_upstairs = self.active_upstairs.get(&upstairs_uuid).unwrap();
+
+        if active_upstairs.upstairs_connection != upstairs_connection {
+            println!(
+                "{:?} cannot grab lock, does not match {:?}!",
+                upstairs_connection, active_upstairs.upstairs_connection,
+            );
+
+            bail!(CrucibleError::UpstairsInactive)
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply downstairs/src/lib.rs:1395]: These should both be returning `UpstairsInactive`, as the Upstairs has to bail either way. Changed it to a print + bail in 89c2384.

File: downstairs/src/lib.rs [hunk 6]
```diff
@@ -1443,8 +1491,13 @@ impl Downstairs {
...
      */
-    async fn complete_work(&mut self, ds_id: u64, m: Message) -> Result<()> {
-        let mut work = self.work.lock().await;
+    async fn complete_work(
+        &mut self,
+        upstairs_connection: UpstairsConnection,
+        ds_id: u64,
+        m: Message,
+    ) -> Result<()> {
+        let mut work = self.work_lock(upstairs_connection).await?;
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply downstairs/src/lib.rs:1506]: You're right, each connection gets a unique work queue.

I tried to address this in aa2b3a5 - there is one scenario I could think of where that could be None, and I think you're right that it shouldn't touch the Work struct. Also added some tests there.

File: downstairs/src/lib.rs [hunk 7]
```diff
@@ -1468,9 +1521,11 @@ impl Downstairs {
      */
     async fn complete_work_stat(
         &mut self,
+        _upstairs_connection: UpstairsConnection,
         m: &Message,
         ds_id: u64,
     ) -> Result<()> {
+        // XXX dss per upstairs connection?
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply downstairs/src/lib.rs:1541]: opened https://github.com/oxidecomputer/crucible/issues/416

File: downstairs/src/lib.rs (hunk 2 above)
Comment (jmpesp) [reply downstairs/src/lib.rs:3555]: Good idea, done in 39514d5

File: downstairs/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (leftwo) [reply downstairs/src/lib.rs:1541]: thanks

File: downstairs/src/lib.rs (hunk 1 above)
Comment (jmpesp) [reply downstairs/src/lib.rs:1648]: I think we're ok here - `promote_to_active` needs a `&mut self`, meaning an exclusive reference, meaning I don't think two calls of this can race.

File: integration_tests/src/lib.rs (hunk 3 above)
Comment (leftwo) [reply integration_tests/src/lib.rs]: Oh, and sorry, but there is some churn in this area with all the scrubber tests I added.

File: downstairs/src/lib.rs [hunk 8]
```diff
@@ -1499,106 +1554,231 @@ impl Downstairs {
...
-         */
-        if work.active.keys().len() > 0 {
-            println!(
-                "Crucible Downstairs promoting {:?} to active, \
-                discarding {} jobs",
-                upstairs_connection,
-                work.active.keys().len()
+            // Insert a new session, overwritting the previous entry if the
+            // Upstairs UUID has an entry already.
+            self.active_upstairs.insert(
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply downstairs/src/lib.rs:1757]: I think we're ok here - `promote_to_active` needs a &mut self, meaning an exclusive reference. I made `work_lock` (and others) require `&mut self` to ensure that callers need an exclusive reference in order to read from or modify `self.active_upstairs`. See: ca64e99

File: downstairs/src/lib.rs [hunk 9]
```diff
@@ -1473,14 +1485,133 @@ impl Downstairs {
...
+            // UpstairsConnection, but the tasks can still be working on
+            // outdated job IDs. If that happens, `get_ready_job` will return a
+            // None, so bail early here.
+            if job.is_none() {
+                return Ok(None);
+            }
+
+            job.unwrap()
+        };
+
```
Comment (leftwo) [inline downstairs/src/lib.rs:1511]: The old do_work() had a few asserts here:
```
    assert_eq!(job.state, WorkState::InProgress);
    assert_eq!(job_id, job.ds_id);
```
Would those still make sense?

It also had a check to be sure the deps were satisfied.  Same question for that.

Comment (leftwo) [reply downstairs/src/lib.rs:1511]: Ah, nope, nevermind.
I see this now happens in the `get_ready_job()` method.
So, why is this comment even here still?  Because I want you to see how hard I'm working, that's why.

**Review by leftwo — APPROVED** [review]
